containers-tests/tests/map-strictness.hs

```diff
@@ -108,6 +108,20 @@ pStrictFoldlWithKey' :: Map Int Int -> Property
 pStrictFoldlWithKey' m = whnfHasNoThunks (M.foldlWithKey' (\as _ a -> a : as) [] m)
 #endif
 
+#if __GLASGOW_HASKELL__ >= 806
+pStrictFromDistinctAscList :: [Int] -> Property
+pStrictFromDistinctAscList = whnfHasNoThunks . evalSpine . M.elems . M.fromDistinctAscList . zip [0::Int ..] . map (Just $!)
+  where
+    evalSpine xs = length xs `seq` xs
+#endif
```

Comment:
Note to self: Should we make it absolutely clear that those are really thunks going in by using `(Just $!) . hiddenId`, where `hiddenId = id`, but it's defined in an unoptimized module that gets no strictness analysis and no unfoldings? As it is, we seem to be relying on `map (Just $!)` not getting optimized into the `Arbitrary [Int]` generation. I'm almost certain it won't be (with current GHC), but that feels a tad fragile.

Reply:
I had wondered about that, too. However, I don't think the `hiddenId` would solve the problem, because it's not the `Int` that is the thunk, it's the `Maybe`. I found I had to use `Just $!` rather than `Just` otherwise even with the fix there were still thunks in the `Int` that QuickCheck provides. Doing it this way, the thunk is then a `Just` of an already-evaluated `Int`.

Like you say, though, there's nothing to stop a future optimizer deciding that the `Just` doesn't need to be a thunk and inserting it directly. To stop that I think we'd need a `hiddenJust`. Even better, though, would be to use something other than `Maybe`, since that obscures the purpose of what we're trying to do. A possible alternative would be [Box](https://hackage.haskell.org/package/ghc-lib-parser/docs/GHC-Exts-Heap.html#t:Box) from `GHC.Exts.Heap`. That's designed to hold bottoms safely, so I don't think it would ever be optimized away.

Reply:
I'm not totally understanding what you said there. The point is that when GHC sees "Store `Just $! x` in the heap" (in this case, in a list cons), what code it generate depends on whether it knows that `x` is already in WHNF. If it does, then it can just drop the `$!`, so the code will allocate a `Just` object. If it doesn't know that, then it'll instead allocate a thunk that, when forced, will evaluate `x` and store a pointer to the result in a `Just` object. We want the latter behavior. You're right that `Just` is the wrong shape, but `Box` is overkill—we don't need its pseudo-existential type. The right tool is `Data.Tuple.Solo`. The place to get that for testing (for now) is the `OneTuple` compatibility package.

Reply:
`(Just $!) . hiddenId` does the trick because it makes sure GHC doesn't know that the argument to `Just $!` is already in WHNF. The same goes for `MkSolo` rather than `Just`.

Reply:
Yes, I see what you mean. That's a more precise analysis than I had.

Yes, I think `Solo` would be good. I had considered it but forgot about it when I made my previous comment.